File: src/main/kotlin/creational/prototype/java/Application.java
package creational.prototype.java;

public class Application {
    public static void main(String... args) {
        Car basicRenault = new Car("0982-9874-9823", "Renault", "Red");
        basicRenault.addFeature("Feature 1");

        Car lessBasicRenault = (Car) basicRenault.clone();
        lessBasicRenault.addFeature("Feature 2");

        Car badassRenault = (Car) lessBasicRenault.clone();
        badassRenault.addFeature("Feature 3");
        badassRenault.addFeature("Feature 4");

        System.out.println("Original Car:");
        System.out.println("Model: " + basicRenault.getModel());
        System.out.println("Color: " + basicRenault.getColor());
        System.out.println("Features: " + basicRenault.getFeatures());

        System.out.println("Cloned Car 1 (less basic):");
        System.out.println("Model: " + lessBasicRenault.getModel());
        System.out.println("Color: " + lessBasicRenault.getColor());
        System.out.println("Features: " + lessBasicRenault.getFeatures());

        System.out.println("Cloned Car 2 (badass):");
        System.out.println("Model: " + badassRenault.getModel());
        System.out.println("Color: " + badassRenault.getColor());
        System.out.println("Features: " + badassRenault.getFeatures());
    }
}

File: src/main/kotlin/creational/prototype/kotlin/Application.kt
package creational.prototype.kotlin

fun main() {
    val basicCar = Car(
        innerCarNumber = "0983-9873-0987",
        model = "Renault",
        color = "Red",
        features = listOf("Feature 1")
    )

    val lessBasicCar = basicCar.copy(features = basicCar.features.plus("Feature 2"))
    val badassCar = basicCar.copy(features = basicCar
        .features
        .plus("Feature 3")
        .plus("Feature 4")
    )

    println("Original Car:");
    println("Model: " + basicCar.model);
    println("Color: " + basicCar.color);
    println("Features: " + basicCar.features);

    println("Cloned Car 1 (less basic):");
    println("Model: " + lessBasicCar.model);
    println("Color: " + lessBasicCar.color);
    println("Features: " + lessBasicCar.features);

    println("Cloned Car 2 (badass):");
    println("Model: " + badassCar.model);
    println("Color: " + basicCar.color);
    println("Features: " + badassCar.features);
}
